### ISSUE TITLE
ability to kill (long-running) pipeline

### ISSUE THREAD

**Issue by Phaqui** [rust-shell-script/rust_cmd_lib#55]
My use-case is that I have a long-running pipeline that runs on web request. If the connection is dropped before the processing is finished, the pipeline still continues to run until completion. If the handle returned from `spawn!()` had a `kill()` method on it, I could call it in the appropriate place for the web framework to stop the pipeline when the connection was dropped. I'm thinking something akin to `std::process::Child::kill()`, maybe on the `CmdChildren` struct.

Or is there maybe an alternative way I can achieve the same without having to implement new functionality?

**Comment by rust-shell-script** [comment]
Just pushed a change. Could you verify if it solved your problem ?

**Comment by Phaqui** [comment]
Hey, thank you for a very quick response - and a patch, too! I did not expect it this fast! :)

To be honest I am not able to work up some code to actually test whether or not what you implemented will work for me, or not. My initial point of struggle, is that both `wait()` and `kill()` takes `&mut self`, and I can't seem to find a way around that. I sort of want two references to the `CmdChildren` (or actually `FunChildren`, because I want the output of the pipeline that I'm running - although as a workaround, I can probably just `tee` the final output to a file, so that I can use `CmdChildren`, which has `kill()`).

I have one task that creates the pipeline, and then blocks on `wait()`. Additionally, I want another task that also has a reference, which waits for an "abort signal", and if that signal comes, then can call `kill()`. The straight forward code quickly leads to a "use after move". Wrapping it in an `Arc<Mutex<_>>` gets me around that, but the first task will get to `lock()` it first, and so even if the code compiles, the second task will be blocked on the mutex until the `.wait()-ing` task is done with the lock - which is the entire problem I'm trying to solve, hehe.

I have been experimenting with other strategies, such as `FunChildren::wait_with_pipe()`, but haven't gotten anywhere with that either.

I might have to put this issue on hold. It doesn't block my application from working, it's just a little bit..unideal. If that's even a word.

**Comment by rust-shell-script** [comment]
I see, in that case I have added API to get the list of pids and you can send kill signal to them: https://docs.rs/nix/latest/nix/sys/signal/fn.kill.html.

**Comment by Phaqui** [comment]
Yes, this works [like a charm](https://github.com/giellatekno/webpipelines/commit/aa05df1ad6ca80de2381804602f939d9a02f4ffa#diff-72751f4d51fd654cc146877a8c74972b97ed12e29c272e3e5fcc78f50436109fR81-R147)!

Thank you so much!
